analyzers/src/SonarAnalyzer.CSharp/Rules/TimestampsShouldNotBeUsedAsPrimaryKeys.cs

```diff
@@ -0,0 +1,69 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class TimestampsShouldNotBeUsedAsPrimaryKeys : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S3363";
+    private const string MessageFormat = "Timestamps should not be used as primary keys";
+
+    private static readonly KnownType[] TemporalTypes = new[]
+    {
+        KnownType.System_DateTime,
+        KnownType.System_DateTimeOffset,
+        KnownType.System_TimeSpan
+    };
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
+        {
+            var classDeclaration = (ClassDeclarationSyntax)c.Node;
+            var className = classDeclaration.Identifier.ValueText;
+            var keyProperties = classDeclaration.Members
```

Comment:
Do you think it would decrease the readability if you would remove the `keyProperties` variable?
It is used only once, so it is not really bringing any value.

Reply:
I find it more readable, because the expression behind it is complex: a property access and two LINQ method calls.
I could add it inside the `foreach` loop, but for me that's more cognitive load than keeping it as a separate variable.

```
foreach (var keyProperty in classDeclaration.Members
								.OfType<PropertyDeclarationSyntax>()
								.Where(x => IsTemporalType(x) && IsKeyProperty(x, className)))
{
	c.ReportIssue(Diagnostic.Create(Rule, keyProperty.Type.GetLocation()));
}
```